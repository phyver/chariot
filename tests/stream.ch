data nat where Zero : nat | Succ : nat -> nat

val add : nat -> nat -> nat
    | add n 0 = n
    | add n (m+1) = Succ (add n m)

val mult : nat -> nat -> nat
    | mult n 0 = 0
    | mult n (m+1) = n + (mult n m)

val pow : nat -> nat -> nat
    | pow n 0 = 1
    | pow n (m+1) = mult n (pow n m)


codata stream('x) where
    | Head : stream('x) -> 'x
    | Tail : stream('x) -> stream('x)

val map : ('x -> 'y) -> stream('x) -> stream('y)
    | (map f s).Head = f (s.Head)
    | (map f s).Tail = map f s.Tail

val arith : nat -> nat -> stream(nat)
  | (arith n r).Head = n
  | (arith n r).Tail = arith (n+r) r

val get_nth : nat -> stream('a) -> 'a
  | get_nth Zero s = s.Head
  | get_nth (Succ n) s = get_nth n s.Tail

val double : nat -> nat | double n = mult n 2

:reduce map double (arith 3 3)

:reduce get_nth (mult 3 3) (map double (arith 3 3))


val (fib n m).Head = n
  | (fib n m).Tail = fib m (n+m)


:reduce get_nth (4+7) (map double (fib 0 1))

