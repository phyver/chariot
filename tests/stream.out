>> inductive type nat was succesfully defined
>> Typing for add successful
>> the definition for add is complete
>> the definition add is provably correct

>> Typing for mult successful
>> the definition for mult is complete
>> the definition mult is provably correct

>> Typing for pow successful
>> the definition for pow is complete
>> the definition pow is provably correct

>> coinductive type stream was succesfully defined
>> Typing for map successful
>> the definition for map is complete
>> the definition map is provably correct

>> Typing for arith successful
>> the definition for arith is complete
>> the definition arith is provably correct

>> Typing for get_nth successful
>> the definition for get_nth is complete
>> the definition get_nth is provably correct

>> Typing for double successful
>> the definition for double is complete
>> the definition double is provably correct

>> term: map double (arith 3 3)
>> result: map double (arith 3 3)
>> of type: stream(nat)

>> term: get_nth (mult 3 3) (map double (arith 3 3))
>> result: 60
>> of type: nat

>> Typing for fib successful
>> the definition for fib is complete
>> the definition fib is provably correct

>> term: get_nth 11 (map double (fib 0 1))
>> result: 178
>> of type: nat

